### PR TITLE
Memoize UnitBase._get_physical_type_id

### LINKED ISSUE #13495: UnitBase._get_physical_type_id is expensive and could be memoized
### Description

Profiling units code, much time is spent in `Unit._get_physical_type_id`, which is used in e.g. `is_equivalent`.

Although this never changes per unit instance, this does expensive operations every time it is called (decompose, list comprehension, conversion to tuple.

**Describe the solution you'd like**

Use caching (replace with lazyproperty) or calculate once in `__init__`.

**Describe alternatives you've considered**
None, this is a major bottleneck.

**Additional context**

Profiling creation of `Angle`, `Latitude`, `Longitudes`:

```
import astropy.units as u
from astropy.units import Quantity
import numpy as np
from astropy.coordinates.angles import Longitude, Latitude, Angle

values = np.random.uniform(-89, 89, 100)

for cls in ('Longitude', 'Latitude', 'Angle', 'Quantity'):
    print(cls)
    get_ipython().run_line_magic('timeit', f'{cls}(values, u.deg)')
```

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  1865738    3.235    0.000   18.422    0.000 /home/maxnoe/Projects/astropy/astropy/units/quantity.py:407(__new__)
  1950033    2.698    0.000    5.314    0.000 /home/maxnoe/Projects/astropy/astropy/units/core.py:665(_get_physical_type_id)
  1054469    2.398    0.000   19.349    0.000 /home/maxnoe/Projects/astropy/astropy/coordinates/angles.py:107(__new__)
    81118    1.967    0.000    5.951    0.000 /home/maxnoe/Projects/astropy/astropy/coordinates/angles.py:380(_wrap_at)
  1947225    1.188    0.000    1.189    0.000 {built-in method numpy.array}
   973371    1.131    0.000    9.220    0.000 /home/maxnoe/Projects/astropy/astropy/units/core.py:916(is_equivalent)
 12034225    1.063    0.000    1.064    0.000 {built-in method builtins.isinstance}
2109569/2109567    0.894    0.000    1.625    0.000 {method 'view' of 'numpy.ndarray' objects}
```

Screenshot_20220725_151212

### Expected behavior

Memoize this, as it is done for `hash`.

### Actual behavior

Expensive operations are performed each time.

### Steps to Reproduce

See benchmark code above.

### PR DESCRIPTION
**Pull request by maxnoe**
### Description

This pull request is to address ...

Adds the same memoization strategy as for `_hash` for `_type_id` used in `_get_physical_type_id`

Fixes #13495 

Timings:

master:

```
Longitude
66.3 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Latitude
34.6 µs ± 479 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Angle
7.44 µs ± 52.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Quantity
1.75 µs ± 14.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

This branch:

```
Longitude
63.5 µs ± 1.47 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Latitude
30.4 µs ± 225 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Angle
5.21 µs ± 15.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Quantity
1.79 µs ± 26.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/units/core.py [hunk 1]
```diff
@@ -669,12 +670,14 @@ def _get_physical_type_id(self):
...
-        # bases and powers are already sorted in a unique way
-        # r.sort()
-        r = tuple(r)
-        return r
+        if self._type_id is None:
+            unit = self.decompose()
+            r = zip([x.name for x in unit.bases], unit.powers)
+            # bases and powers are already sorted in a unique way
+            # r.sort()
+            self._type_id = tuple(r)
```
Comment (eerovaher) [inline astropy/units/core.py:678]: This code will only be executed if `_type_id` isn't computed already, but replacing the list comprehension with a generator expression should be faster.
```suggestion
            self._type_id = tuple(zip((base.name for base in unit.bases), unit.powers))
```

**Comment by maxnoe** [comment]
@mhvk done

**Comment by maxnoe** [comment]
I think a changelog entry is not necessary for each of these micro optimizations, but we should probably include a common what's new? Or how do you want to handle this?

**Review by mhvk — APPROVED** [review]
Super! I'll get this in, but agree that for one of those angle-related PRs, we should add a changelog entry. Or we can just do a general what's new entry with improvements in units and angles.

**Pull request merged**